### PR TITLE
Create dicom uploader function

### PR DESCRIPTION
**Pull request by brandonpollett**
## Description
This PR adds a dicom uploader function to the tools directory for using an Azure Function with a blob trigger to do uploads to a configured DICOM Service.

## Related issues
Addresses [AB#92097](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/92097).

## Testing
Manually tested.

### CODE REVIEW

File: converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.Development.json [hunk 1]
```diff
@@ -19,7 +19,15 @@
   "Fhir": {
     "Endpoint": "https://localhost:44348",
     "Authentication": {
-      "Enabled": false
+      "Enabled": true,
```
Comment (wsugarman) [inline converter/dicom-cast/src/Microsoft.Health.DicomCast.Hosting/appsettings.Development.json:22]: Will we need an update for this in the docker-compose as well?

File: tools/uploader-function/.gitignore [hunk 2]
```diff
@@ -0,0 +1,350 @@
+## Ignore Visual Studio temporary files, build results, and
```
Comment (wsugarman) [inline tools/uploader-function/.gitignore:1]: Do we need another `.gitignore`?

File: tools/uploader-function/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Uploader Function
+
+This solution uses the Azure Function framework to allow uploads to a DICOM Service.
```
Comment (wsugarman) [inline tools/uploader-function/README.md:3]: Should we have a link to Azure functions or the az cli module?

File: tools/uploader-function/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+## Create function app
+```
+az functionapp create --name dicom-uploader --storage-account {storage-account-name} --resource-group {resource-group-name} --plan {app-service-plan-name} --deployment-container-image-name dicomoss.azurecr.io/dicom-uploader:0.0.1 --functions-version 4 --assign-identity [system]
+```
+
+## Grant permissions for function's managed identity
+
+- Grant Function managed identity Dicom Data Owner
+- Grant Function managed identity Storage Blob Contrib
+- Grant Function Managed identity Storage Queeu Contrib
```
Comment (wsugarman) [inline tools/uploader-function/README.md:20]:
```suggestion
- Grant Function Managed identity Storage Queue Contributor
```

File: tools/uploader-function/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+## Create function app
+```
+az functionapp create --name dicom-uploader --storage-account {storage-account-name} --resource-group {resource-group-name} --plan {app-service-plan-name} --deployment-container-image-name dicomoss.azurecr.io/dicom-uploader:0.0.1 --functions-version 4 --assign-identity [system]
+```
+
+## Grant permissions for function's managed identity
+
+- Grant Function managed identity Dicom Data Owner
+- Grant Function managed identity Storage Blob Contrib
```
Comment (wsugarman) [inline tools/uploader-function/README.md:19]:
```suggestion
- Grant Function managed identity Storage Blob Contributor
```

File: tools/uploader-function/src/DicomUploaderFunction/.gitignore [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# Azure Functions localsettings file
+local.settings.json
```
Comment (wsugarman) [inline tools/uploader-function/src/DicomUploaderFunction/.gitignore:2]: Is this file necessary? I think we already ignore this

File: tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace DicomUploaderFunction;
+
+public class BlobUploader
+{
+    private readonly IDicomWebClient _dicomWebClient;
+
+    public BlobUploader(IDicomWebClient dicomWebClient)
+    {
+        _dicomWebClient = dicomWebClient;
```
Comment (wsugarman) [inline tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs:23]: nit: `null`-check?

File: tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    [FunctionName("BlobUploader")]
+    public async Task Run(
+        [BlobTrigger("%sourceblobcontainer%/{name}", Connection = "sourcestorage")]
+        Stream myBlob,
+        string name,
```
Comment (wsugarman) [inline tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs:30]: Is `name` needed?

File: tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        ILogger log)
+    {
+        using var streamContent = new StreamContent(myBlob);
+        streamContent.Headers.ContentType = new MediaTypeHeaderValue("application/dicom");
+        await _dicomWebClient.StoreAsync(streamContent, partitionName: null, CancellationToken.None);
```
Comment (wsugarman) [inline tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs:35]: I think you can add a token parameter as well that will cancel when the functions host shuts down

File: tools/uploader-function/src/DicomUploaderFunction/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    mkdir -p /home/site/wwwroot && \
+    dotnet publish *.csproj --output /home/site/wwwroot
+
+# To enable ssh & remote debugging on app service change the base image to the one below
+# FROM mcr.microsoft.com/azure-functions/dotnet:4-appservice
+FROM mcr.microsoft.com/azure-functions/dotnet:4.6.1-slim@sha256:11bc18a02197d460dfa96fe2aabde5a20606c85892bb670224e92c20e045ec0b 
+ENV AzureWebJobsScriptRoot=/home/site/wwwroot \
+    AzureFunctionsJobHost__Logging__Console__IsEnabled=true
+
+COPY --from=installer-env ["/home/site/wwwroot", "/home/site/wwwroot"]
```
Comment (wsugarman) [inline tools/uploader-function/src/DicomUploaderFunction/Dockerfile:16]: nit: newline after last line

File: tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    private readonly IDicomWebClient _dicomWebClient;
+
+    public BlobUploader(IDicomWebClient dicomWebClient)
+    {
+        _dicomWebClient = dicomWebClient;
+    }
+
+    [FunctionName("BlobUploader")]
+    public async Task Run(
+        [BlobTrigger("%sourceblobcontainer%/{name}", Connection = "sourcestorage")]
```
Comment (wsugarman) [inline tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs:28]: Should this value be scoped to some section? Eg. `"%foo:bar:baz%"`?

File: tools/uploader-function/README.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
+# Uploader Function
+
+This solution uses the Azure Function framework to allow uploads to a DICOM Service.
+
+Steps:
+## Create app service plan
+```
```
Comment (mmitrik) [inline tools/uploader-function/README.md:7]: We talked about some of this on Friday, but here it would be good to call out detail about these parameters.

For this one, I think the resource group needs to be the same one running the DICOM service.

File: tools/uploader-function/README.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+az functionapp plan create --resource-group {resource-group-name} --name {app-service-plan-name} --location {REGION} --number-of-workers 1 --sku EP1 --is-linux
+```
+
+## Create function app
+```
```
Comment (mmitrik) [inline tools/uploader-function/README.md:12]: Here call out:
- Same `{resource-group}` as above
- Same `{app-service-plan-name}` as above
- `{storage-account-name}` is the storage account that the function app will use during the import, NOT the source storage account
- Its safe to ignore warnings about lack of credentials accessing the container registry (it's publicly accessible)

File: tools/uploader-function/README.md (hunk 5 above)
Comment (mmitrik) [reply tools/uploader-function/README.md:19]: In line with my prior note, this is the function app's storage account

File: tools/uploader-function/README.md [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+## Grant permissions for function's managed identity
+
+- Grant Function managed identity Dicom Data Owner
+- Grant Function managed identity Storage Blob Contrib
+- Grant Function Managed identity Storage Queeu Contrib
+
+## Update settings
+```
```
Comment (mmitrik) [inline tools/uploader-function/README.md:23]: Here, I'd note that the `sourceblobcontainer` is not the one from above, it's where the DICOM data currently lives and will be migrated from.

File: tools/uploader-function/README.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+az functionapp config appsettings set --name dicom-uploader --resource-group {resource-group-name} --settings "sourceblobcontainer={source-container-name}" "sourcestorage__blobServiceUri={source-blob-url}" "sourcestorage__queueServiceUri={source-queue-url}" "DicomWeb__Endpoint={dicom-service-endpoint}" "DicomWeb__Authentication__Enabled=true" "DicomWeb__Authentication__AuthenticationType=ManagedIdentity" "DicomWeb__Authentication__ManagedIdentityCredential__Resource=https://dicom.healthcareapis.azure.com"
+```
```
Comment (mmitrik) [inline tools/uploader-function/README.md:25]: Some other items we discussed on Friday 
- What happens if you need to re-run to import from another storage container?  You should just need to update these app settings - but does the function need stopped?  Any other gotchas?
- A pointer for troubleshooting, even just a pointer to the poison queues folder would be good.  
- If a customer was running this and contacted us saying they didn't see anything migrated, what is our first 1-2 question we would ask them?  Can we pre-emptively include something here?

**Review by mmitrik — COMMENTED** [review]
Only looked at the docs, but left a few comments there from the run-through on Friday.

File: tools/uploader-function/src/DicomUploaderFunction/.gitignore (hunk 6 above)
Comment (brandonpollett) [reply tools/uploader-function/src/DicomUploaderFunction/.gitignore:2]: Doesn't look like we do, but I'll add it to the global `.gitignore`

File: tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs (hunk 11 above)
Comment (brandonpollett) [reply tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs:28]: It's an environment variable so we should be able to, but we'd probably need to use `__` instead of `:`. We'd likely set it to somthing like `sourcestorage__conainer`

**Review by mmitrik — DISMISSED** [review]
Found a few typos.  Otherwise, the doc changes look great :D

File: .gitignore [hunk 16]
```diff
@@ -290,3 +290,6 @@ tempkey.rsa
 
 # docker .env file
 **/.env
+
+# Local settings files
+local.settings.json
```
Comment (wsugarman) [inline .gitignore:295]: I think we want this in the `.gitignore` for the F5 scenario

File: tools/uploader-function/README.md [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+This step creates a Linux premium app service plan that is able to host a container function.
+
+```
+az functionapp plan create --resource-group {resource-group-name} --name {app-service-plan-name} --location {REGION} --number-of-workers 1 --sku EP1 --is-linux
+```
+
+| Property | Value |
+| --- | --- |
+| {resource-group-name} | The resource group that you would like to place your app service plan in. |
+| {app-service-plan-name} | The name of your app service plan.
```
Comment (wsugarman) [inline tools/uploader-function/README.md:16]: nit: missing trailing pipe `|`

File: tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs (hunk 11 above)
Comment (wsugarman) [reply tools/uploader-function/src/DicomUploaderFunction/BlobUploader.cs:28]: I use this in [our own functions](https://github.com/microsoft/healthcare-shared-components/blob/f5ef8dd8c2fe80484f44754fd6b3c26cd17e9225/src/Microsoft.Health.Operations.Functions/Management/PurgeOrchestrationInstanceHistory.cs#L27) and the syntax appears to be `:` like configs (I guess to normalize it?)

File: tools/uploader-function/src/DicomUploaderFunction/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+COPY ./tools/uploader-function/src/DicomUploaderFunction /tools/uploader-function/src/DicomUploaderFunction
+RUN cd /tools/uploader-function/src/DicomUploaderFunction && \
+    mkdir -p /home/site/wwwroot && \
+    dotnet publish *.csproj --output /home/site/wwwroot
+
+# To enable ssh & remote debugging on app service change the base image to the one below
+# FROM mcr.microsoft.com/azure-functions/dotnet:4-appservice
+FROM mcr.microsoft.com/azure-functions/dotnet:4.6.1-slim@sha256:11bc18a02197d460dfa96fe2aabde5a20606c85892bb670224e92c20e045ec0b 
+ENV AzureWebJobsScriptRoot=/home/site/wwwroot \
+    AzureFunctionsJobHost__Logging__Console__IsEnabled=true
```
Comment (wsugarman) [inline tools/uploader-function/src/DicomUploaderFunction/Dockerfile:14]: You may not need console logging if it's hosted in an app service, since it will presumably output to app insights

**Review by wsugarman — APPROVED** [review]
